File: examples/python/controller_usage_example.py
#!/usr/bin/env python

import rospy
import time
import numpy as np
from std_msgs.msg import Float32MultiArray
from std_msgs.msg import Float64

class Robot():
    def __init__(self):
        self.dof = 6
        self.robot_goal_pub = rospy.Publisher('/fanuc_stmotion_controller_bringup/robot_goal', Float32MultiArray, queue_size=6)
        self.robot_state_sub = rospy.Subscriber("/fanuc_stmotion_controller_bringup/robot_state", Float32MultiArray, self.robot_state_callback)
        self.robot_travel_time_pub = rospy.Publisher('/fanuc_stmotion_controller_bringup/jpc_travel_time', Float64, queue_size=1)
        self.robot_goal = np.zeros(self.dof)
        self.robot_goal_msg = Float32MultiArray()
        self.robot_state = np.zeros(self.dof)
        self.robot_state_v = np.zeros(self.dof)
        self.robot_state_a = np.zeros(self.dof)
        self.robot_travel_time = Float64()

    def robot_state_callback(self, data):
        for i in range(self.dof):
            self.robot_state[i] = data.data[i*3]
            self.robot_state_v[i] = data.data[i*3 + 1]
            self.robot_state_a[i] = data.data[i*3 + 2]
    
    def set_robot_travel_time(self, t):
        self.robot_travel_time.data = t
        self.robot_travel_time_pub.publish(self.robot_travel_time)

    def drive_robot(self, goal):
        for i in range(self.dof):
            self.robot_goal[i] = goal[i]
        self.robot_goal_msg.data = goal
        self.robot_goal_pub.publish(self.robot_goal_msg)

    def goal_reached(self):
        for i in range(self.dof):
            if(abs(self.robot_state[i] - self.robot_goal[i]) > 0.01 or abs(self.robot_state_v[i]) > 0.001 or abs(self.robot_state_a[i] > 0.001)):
                return False
        return True

def main():
    rospy.init_node('use_controller_node')
    fanuc = Robot()
    goal_list = [[0, 0, 0, 0, 0, 0],
                 [0, 0, 0, 0, -90, 0],
                 [0, 0, 0, 0, 0, 0],
                 [0, 0, 0, 0, -90, 0],
                 [0 ,0, 0, 0, 0, 0]]
    
    goal_idx = -1
    cur_goal = goal_list[0]
    time.sleep(0.1)
    while not rospy.is_shutdown():
        if(goal_idx == -1 or fanuc.goal_reached()):
            print("Reached: ", goal_idx)
            if(goal_idx == -1):
                fanuc.set_robot_travel_time(0.5)
            elif(goal_idx == 1):
                fanuc.set_robot_travel_time(5)
            elif(goal_idx == 3):
                fanuc.set_robot_travel_time(0.1)
            
            goal_idx += 1
            if(goal_idx < len(goal_list)):
                cur_goal = goal_list[goal_idx]   
            else:
                break
            time.sleep(0.01)
        fanuc.drive_robot(cur_goal)
            
if __name__ == '__main__':
    main()

File: examples/cpp/controller_usage_example.cpp
#include "Robot.hpp"

using namespace std::chrono;

stmotion_controller::math::VectorJd robot_q = Eigen::MatrixXd::Zero(6, 1);
stmotion_controller::math::VectorJd robot_qd = Eigen::MatrixXd::Zero(6, 1);
stmotion_controller::math::VectorJd robot_qdd = Eigen::MatrixXd::Zero(6, 1);

void robotStateCallback(const std_msgs::Float32MultiArray::ConstPtr& msg)
{
    robot_q << msg->data[0], msg->data[3], msg->data[6], msg->data[9], msg->data[12], msg->data[15];
    robot_qd << msg->data[1], msg->data[4], msg->data[7], msg->data[10], msg->data[13], msg->data[16];
    robot_qdd << msg->data[2], msg->data[5], msg->data[8], msg->data[11], msg->data[14], msg->data[17];
}

bool reached_goal(const stmotion_controller::math::VectorJd& goal)
{
    for(int i=0; i<6; i++)
    {
        if(abs(goal(i) - robot_q(i)) > 0.01 || abs(robot_qd(i)) > 0.001 || abs(robot_qdd(i)) > 0.001)
        {
            return false;
        }
    }
    return true;
}

int main(int argc, char **argv)
{
    try
    {
        ros::init(argc, argv, "use_controller_node");
        ros::NodeHandle nh("~");
        ROS_INFO_STREAM("namespace of nh = " << nh.getNamespace());
        ros::Rate loop_rate(150);

        Eigen::Matrix<double, 6, 5> joint_goal = Eigen::Matrix<double, 6, 5>::Zero();
        joint_goal.col(0) << 0, 0, 0, 0, 0, 0;
        joint_goal.col(1) << 0, 0, 0, 0, -90, 0;
        joint_goal.col(2) << 0, 0, 0, 0, 0, 0;
        joint_goal.col(3) << 0, 0, 0, 0, -90, 0;
        joint_goal.col(4) << 0, 0, 0, 0, 0, 0;
        int robot_dof = 6;
       
        ros::Publisher goal_pub = nh.advertise<std_msgs::Float32MultiArray>("/stmotion_controller_bringup/robot_goal", robot_dof);
        ros::Publisher jpc_time_pub = nh.advertise<std_msgs::Float64>("/stmotion_controller_bringup/jpc_travel_time", 1);
        ros::Subscriber robot_state_sub = nh.subscribe("/stmotion_controller_bringup/robot_state", robot_dof * 3, robotStateCallback);
        std_msgs::Float32MultiArray goal_msg;
        std_msgs::Float64 jpc_time_msg;
        
        int num_tasks = joint_goal.cols();
        int task_idx = -1;
        stmotion_controller::math::VectorJd cur_goal = Eigen::MatrixXd::Zero(6, 1);
        cur_goal = joint_goal.col(0);
        unsigned int microsecond = 1000;
        usleep(300 * microsecond);
        while(ros::ok)
        {
            if(task_idx == -1 || reached_goal(cur_goal))
            {
                if(task_idx == -1)
                {
                    jpc_time_msg.data = 0.5;
                    jpc_time_pub.publish(jpc_time_msg);
                }
                else if (task_idx == 1)
                {
                    jpc_time_msg.data = 5;
                    jpc_time_pub.publish(jpc_time_msg);
                }
                else if(task_idx == 3)
                {
                    jpc_time_msg.data = 0.1;
                    jpc_time_pub.publish(jpc_time_msg);
                }
                task_idx ++;
                if(task_idx >= joint_goal.cols())
                {
                    break;
                }
                cur_goal = joint_goal.col(task_idx);     
            }
            goal_msg.data.clear();
            for(int j=0; j<robot_dof; j++)
            {
                goal_msg.data.push_back(cur_goal(j));
            }
            goal_pub.publish(goal_msg);
            ros::spinOnce();
        }
        ROS_INFO_STREAM("Task Execution Done!");
        ros::shutdown();
        return 0;
    }
    catch(const std::exception& e)
    {
        std::cerr << e.what() << '\n';
    }
}


